### PR TITLE
Fixed a bug which caused a deadlock in `trigger(for:)`.

### PR DESCRIPTION
**Pull request by andersio**
... a programmer mistake. Kinda side effect of `id` as `Any`.

### CODE REVIEW

**Comment by sharplet** [comment]
Nice.

Do you think it would be worth adding a private/internal `synchronized` helper for `Reactive`, to help avoid this happening with future extensions? Something like

```swift
extension Reactive {
  func synchronized(body: () throws -> Void) rethrows {
    objc_sync_enter(base)
    defer { objc_sync_exit(base) }
    try body()
  }
}
```

Then use `synchronized { }` wherever we currently use `objc_sync_*` directly.

**Comment by sharplet** [comment]
This helper would probably need to be able to return a value, too.

**Comment by andersio** [comment]
BTW, I attempted to write test cases to reproduce #3264 but failed. 🙈

**Comment by andersio** [comment]
@sharplet I constrained it against `NSObject` directly, since it very likely won't be our public API.

**Comment by sharplet** [comment]
I think it would be better to use `synchronized` instead of `synchronize`, in keeping with the `@synchronized` keyword in Onjectove-C.

File: ReactiveCocoa/NSObject+Synchronize.swift [hunk 1]
```diff
@@ -0,0 +1,7 @@
+extension NSObject {
+	internal func synchronized<Result>(execute: () throws -> Result) rethrows -> Result {
```
Comment (NachoSoto) [inline ReactiveCocoa/NSObject+Synchronize.swift]: This can be `final`.

File: ReactiveCocoaTests/InterceptingSpec.swift [hunk 2]
```diff
@@ -71,6 +71,24 @@ class InterceptingSpec: QuickSpec {
 				expect(counter1) == 2
 				expect(counter2) == 2
 			}
+
+			it("should not deadlock") {
```
Comment (NachoSoto) [inline ReactiveCocoaTests/InterceptingSpec.swift]: Awesome

File: ReactiveCocoaTests/LifetimeSpec.swift [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+			}
+
+			afterEach {
+				object = nil
+				expect(_object).to(beNil())
+			}
+
+			it("should not deadlock") {
+				for _ in 1 ... 10 {
+					var isDeadlock = true
```
Comment (NachoSoto) [inline ReactiveCocoaTests/LifetimeSpec.swift]: I'd say `isDeadlocked`.

**Pull request merged**